### PR TITLE
feat(lambda-event-sources) add DynamoEventSource

### LINKED ISSUE #1562: Support for lambda Dynamo event source
Add support for Dynamo event source mappings: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html

### PR DESCRIPTION
**Pull request by sam-goodwin**
Fixes #1562
----
add dynamo lambda event source for consuming DynamoDB streams from AWS Lambda

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by eladb** [comment]
Please correct PR title and add message in accordance with the guidelines.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 1]
```diff
@@ -442,6 +463,16 @@ export class Table extends Construct {
     this.grant(principal, ...READ_DATA_ACTIONS);
   }
 
+  /**
+   * Permis an IAM principal all stream data read operations for this
+   * table's stream:
+   * DescribeStream, GetRecords, GetShardIterator, ListStreams.
+   * @param principal The principal to grant access to
+   */
+  public grantReadStreamData(principal?: iam.IPrincipal) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts]: I guess this should be grantStreamReadData (do you think “data” is really needed?)

File: packages/@aws-cdk/aws-lambda-event-sources/lib/dynamodb.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  public bind(target: lambda.FunctionBase) {
+    new lambda.EventSourceMapping(target, `DynamoDBEventSource:${this.table.node.uniqueId}`, {
+      target,
+      batchSize: this.props.batchSize || 100,
+      eventSourceArn: this.table.tableStreamArn,
+      startingPosition: this.props.startingPosition
+    });
+
+    this.table.grantReadStreamData(target.role);
+    target.addToRolePolicy(new iam.PolicyStatement()
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/dynamodb.ts]: Should we include listStreams in the grant or add another grant method?

Comment (sam-goodwin) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/dynamodb.ts]: I wasn't sure on this one - it doesn't make sense to reference a table instance to grant list streams. Maybe a static method?

**Pull request merged**